ros/src/system/autoware_health_checker/src/diag_buffer.cpp

```diff
@@ -0,0 +1,93 @@
+#include <autoware_health_checker/diag_buffer.h>
+
+namespace autoware_health_checker
+{
+    DiagBuffer::DiagBuffer(std::string key, uint8_t type, std::string description, double buffer_length) : type(type), description(description)
+    {
+        key_ = key;
+        buffer_length_ = ros::Duration(buffer_length);
+    }
+
+    DiagBuffer::~DiagBuffer()
+    {
+        
+    }
+
+    void DiagBuffer::addDiag(autoware_system_msgs::DiagnosticStatus status)
+    {
+        std::lock_guard<std::mutex> lock(mtx_);
+        buffer_[status.level].status.push_back(status);
+        updateBuffer();
+        return;
+    }
+
+    autoware_system_msgs::DiagnosticStatusArray DiagBuffer::getAndClearData()
+    {
+        autoware_system_msgs::DiagnosticStatusArray data;
+        data = buffer_[autoware_health_checker::LEVEL_FATAL];
+        data.status.insert(data.status.end(),buffer_[autoware_health_checker::LEVEL_ERROR].status.begin(),buffer_[autoware_health_checker::LEVEL_ERROR].status.end());
+        data.status.insert(data.status.end(),buffer_[autoware_health_checker::LEVEL_WARN].status.begin(),buffer_[autoware_health_checker::LEVEL_WARN].status.end());
+        data.status.insert(data.status.end(),buffer_[autoware_health_checker::LEVEL_OK].status.begin(),buffer_[autoware_health_checker::LEVEL_OK].status.end());
+        data.status.insert(data.status.end(),buffer_[autoware_health_checker::LEVEL_UNDEFINED].status.begin(),buffer_[autoware_health_checker::LEVEL_UNDEFINED].status.end());
+        std::sort(data.status.begin(), data.status.end(), std::bind(&DiagBuffer::compareTimestamp, this, std::placeholders::_1, std::placeholders::_2));
```

Comment:
I think that the `[ ]` operator of stdmap is not thread-safe.
For element accesses that are not intended to change the value, we recommend accessing `at()` as much as possible.
If it can not be done, it seems necessary to set `mutex` in order to secure thread-safe.

Reply:
add scoped lock to the function.